src/scripts/syncBreaches.js

```diff
@@ -11,7 +11,81 @@
 
 import { req, formatDataClassesArray } from '../utils/hibp.js'
 import { getAllBreaches, upsertBreaches } from '../db/tables/breaches.js'
+import { mkdir, readdir } from "node:fs/promises";
+import { resolve as pathResolve } from "node:path";
+import { finished } from 'node:stream/promises';
+import fs from "node:fs";
+import { Readable } from 'node:stream';
+import aws from 'aws-sdk';
 
+// Get breaches logos and uploads to s3
+const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
+const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
+const region = process.env.AWS_REGION;
+const Bucket = process.env.S3_BUCKET;
+
+const s3 = new aws.S3({
+  region,
+  credentials: {
+    accessKeyId,
+    secretAccessKey,
+  },
+});
+
+async function uploadToS3(fileName, fileStream) {
+  console.log('Attempt to upload to s3: ', fileName)
+  const uploadParams = {
+    Bucket,
+    Key: fileName,
+    Body: fileStream
+  }
+  try {
+    await s3.upload(uploadParams).promise()
+    console.log('Successfully uploaded data to ' + Bucket + '/' + fileName)
+  } catch (err) {
+    console.error(err, err.stack)
+  }
+}
+
+export async function getBreachIcons(breaches) {
+  let logoMap;
+  const breachDomains = breaches
+    .map((breach) => breach.Domain)
+    .filter((breachDomain) => breachDomain.length > 0);
+
+  // make logofolder if it doesn't exist
+  const logoFolder = "public/logo_cache/";
+  if (!fs.existsSync(logoFolder)) await mkdir(logoFolder);
+
+  // read existing logos
+  const existingLogos = await readdir(logoFolder);
+  console.log(`existing logos: ${existingLogos.length}`)
+
+  // TODO: Batch to limit memory use?
+  const logoMapElems = (await Promise.all(
+    breachDomains.map(async (breachDomain) => {
+      const logoFilename = breachDomain.toLowerCase() + ".ico";
+      const logoPath = pathResolve(logoFolder, logoFilename);
+      if (existingLogos.includes(logoFilename)) {
+        console.log('skipping ', logoFilename)
+        return;
+      }
+      console.log(`fetching: , ${logoFilename}`)
+      const res = await fetch(
+        `https://icons.duckduckgo.com/ip3/${breachDomain}.ico`);
+      await uploadToS3(logoFilename, Buffer.from(await res.arrayBuffer()))
+      const fileStream = fs.createWriteStream(logoPath, { flags: 'wx' });
+      const bodyReadable = Readable.fromWeb(res.body)
+      await finished(bodyReadable.pipe(fileStream));
+    })
+  ));
+
+  logoMap = new Map(logoMapElems.filter(e => e != null));
```

Comment:
I don't think we're using `logoMap` anywhere. It seems to be local to this fn (see L51 above) but we aren't returning it from this fn; not that it mattes since the `getBreachIcons()` call on L104 below doesn't expect/use/want a return value.

Reply:
Took it out. thanks for catching this!